crates/env/src/engine/on_chain/pallet_revive.rs

```diff
@@ -1275,52 +1319,125 @@ impl TypedEnvBackend for EnvInstance {
         H256::from_slice(output)
     }
 
-    #[cfg(all(feature = "xcm", feature = "unstable-hostfn"))]
-    fn xcm_execute<E, Call>(&mut self, _msg: &VersionedXcm<Call>) -> Result<()>
+    #[cfg(feature = "xcm")]
+    fn xcm_weigh<Call>(&mut self, msg: &VersionedXcm<Call>) -> Result<Weight>
     where
-        E: Environment,
         Call: scale::Encode,
     {
-        panic!(
-            "todo Native ink! XCM functions are not supported yet, you have to call the pre-compile contracts for XCM directly until then."
-        );
-        /*
         let mut scope = self.scoped_buffer();
+        let enc_msg = scope.take_encoded(msg);
+        let buffer = scope.take_rest();
+
+        let mut output_buffer = [0u8; 64];
+
+        let sel = const { solidity_selector("weighMessage(bytes)") };
+        buffer[..4].copy_from_slice(&sel[..4]);
 
+        let n = solidity_encode_bytes(enc_msg, 32, 0, &mut buffer[4..]);
+
+        let call_result = ext::call(
+            CallFlags::empty(),
+            &XCM_PRECOMPILE_ADDR,
+            u64::MAX,       // `ref_time` to devote for execution. `u64::MAX` = all
+            u64::MAX,       // `proof_size` to devote for execution. `u64::MAX` = all
+            &[u8::MAX; 32], // No deposit limit.
+            &U256::zero().to_little_endian(), // Value transferred to the contract.
+            &buffer[..4 + n],
+            Some(&mut &mut output_buffer[..]),
+        );
+        call_result.expect("call host function failed");
+
+        let mut ref_time_bytes = [0u8; 8];
+        ref_time_bytes.copy_from_slice(&output_buffer[24..32]);
+        let mut proof_size_bytes = [0u8; 8];
+        proof_size_bytes.copy_from_slice(&output_buffer[56..64]);
+
+        let weight: Weight = Weight::from_parts(
+            u64::from_be_bytes(ref_time_bytes),
+            u64::from_be_bytes(proof_size_bytes),
+        );
+        Ok(weight)
+    }
+
+    #[cfg(feature = "xcm")]
+    fn xcm_execute<Call>(
+        &mut self,
+        msg: &VersionedXcm<Call>,
+        weight: Weight,
+    ) -> Result<()>
+    where
+        Call: scale::Encode,
+    {
+        let mut scope = self.scoped_buffer();
         let enc_msg = scope.take_encoded(msg);
+        let buffer = scope.take_rest();
 
-        #[allow(deprecated)]
-        ext::xcm_execute(enc_msg).map_err(Into::into)
-        */
+        let sel = const { solidity_selector("execute(bytes,(uint64,uint64))") };
+        buffer[..4].copy_from_slice(&sel[..4]);
+
+        // 96 because 64 for `Weight` and 32 for `bytes` offset
+        let n = solidity_encode_bytes(enc_msg, 96, 0, &mut buffer[4..]);
+
+        let mut weight_bytes = [0u8; 64];
+        weight_bytes[24..32].copy_from_slice(&weight.ref_time().to_be_bytes());
+        weight_bytes[56..64].copy_from_slice(&weight.proof_size().to_be_bytes());
```

Comment:
nit: can write the weight bytes to `buffer` directly

Reply:
That's a good idea! I've changed it.